File: src/types/agentRuns.ts
/**
 * @file agentRuns.ts
 * @description Agent run audit trail types (v1.5.0 M5.1), mirroring the Rust
 * `AgentRunDto` (camelCase serde). One run = one ReAct loop execution; turns
 * and approvals arrive as the exact JSON documents the auditor persisted.
 */

/** One tool call observed inside a turn, with its execution timing. */
export interface AgentRunToolCall {
  name: string;
  success: boolean;
  durationMs: number;
}

/** One ReAct turn: token usage, wall time, and the tool calls it made. */
export interface AgentRunTurn {
  turn: number;
  promptTokens: number | null;
  completionTokens: number | null;
  durationMs: number;
  toolCalls: AgentRunToolCall[];
}

/** One approval decision inside a run. */
export interface AgentRunApproval {
  turn: number;
  callId: string;
  functionName: string;
  /** "APPROVED" | "REJECTED" | "TIMED_OUT" */
  decision: string;
}

/** Audit trail of a single agent run. */
export interface AgentRunDto {
  id: string;
  conversationId: string;
  branchId: string | null;
  modelId: string | null;
  /** Unix seconds. */
  startedAt: number;
  /** Unix seconds; null while the run is still active. */
  finishedAt: number | null;
  turns: AgentRunTurn[];
  approvals: AgentRunApproval[];
  /** "COMPLETED" | "CANCELLED" | "FAILED"; null while still active. */
  outcome: string | null;
  outcomeDetail: string | null;
}

/** Defensive parse: turns/approvals are JSON documents from the audit table. */
export function asAgentRunTurns(value: unknown): AgentRunTurn[] {
  return Array.isArray(value) ? (value as AgentRunTurn[]) : [];
}

export function asAgentRunApprovals(value: unknown): AgentRunApproval[] {
  return Array.isArray(value) ? (value as AgentRunApproval[]) : [];
}

File: src-tauri/src/dto/agent_runs.rs
/**
 * @file dto/agent_runs.rs
 * @description Agent run audit DTOs (v1.5.0 M5.1).
 *
 * `turns` and `approvals` are carried as pre-built JSON documents (exactly
 * what the auditor persisted) so the frontend can render and export the
 * trail without a second schema to keep in sync.
 */

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::repositories::agent_runs::AgentRunRow;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentRunDto {
    pub id: String,
    pub conversation_id: String,
    pub branch_id: Option<String>,
    pub model_id: Option<String>,
    pub started_at: i64,
    pub finished_at: Option<i64>,
    pub turns: Value,
    pub approvals: Value,
    pub outcome: Option<String>,
    pub outcome_detail: Option<String>,
}

impl AgentRunDto {
    pub fn from_row(row: AgentRunRow) -> Self {
        let parse = |raw: &str| -> Value {
            serde_json::from_str(raw).unwrap_or_else(|_| Value::Array(vec![]))
        };
        Self {
            id: row.id,
            conversation_id: row.conversation_id,
            branch_id: row.branch_id,
            model_id: row.model_id,
            started_at: row.started_at,
            finished_at: row.finished_at,
            turns: parse(&row.turns_json),
            approvals: parse(&row.approvals_json),
            outcome: row.outcome,
            outcome_detail: row.outcome_detail,
        }
    }
}
